slowquant/qiskit_interface/circuit_wavefunction.py

```diff
@@ -1041,262 +947,173 @@ def print_progress_SPSA(
-    return energy
-
-
-def calc_energy_both(parameters, wf) -> float:
-    """Calculate electronic energy.
-
-    Args:
-        parameters: Ansatz and orbital rotation parameters.
-        wf: Wave function object.
-
-    Returns:
-        Electronic energy.
-    """
-    kappa = parameters[: len(wf.kappa)]
-    theta = parameters[len(wf.kappa) :]
-    assert len(theta) == len(wf.ansatz_parameters)
-    # Do orbital partial
-    kappa_mat = np.zeros_like(wf.c_orthonormal)
-    for kappa_val, (p, q) in zip(
-        np.array(kappa) - np.array(wf._kappa_old), wf.kappa_idx  # pylint: disable=protected-access
-    ):
-        kappa_mat[p, q] = kappa_val
-        kappa_mat[q, p] = -kappa_val
-    if len(wf.kappa_redundant) != 0:
-        if np.max(np.abs(wf.kappa_redundant)) > 0.0:
-            for kappa_val, (p, q) in zip(
-                np.array(wf.kappa_redundant)
-                - np.array(wf._kappa_redundant_old),  # pylint: disable=protected-access
-                wf.kappa_redundant_idx,
-            ):
-                kappa_mat[p, q] = kappa_val
-                kappa_mat[q, p] = -kappa_val
-    c_trans = np.matmul(wf.c_orthonormal, scipy.linalg.expm(-kappa_mat))
-    wf._kappa_old = kappa.copy()  # pylint: disable=protected-access
-    wf._kappa_redundant_old = wf.kappa_redundant.copy()  # pylint: disable=protected-access
-    # Moving expansion point of kappa
-    wf.c_orthonormal = c_trans
-    # Build operator
-    wf.ansatz_parameters = theta.copy()  # Reset rdms
-    H = hamiltonian_0i_0a(wf.h_mo, wf.g_mo, wf.num_inactive_orbs, wf.num_active_orbs)
-    H = H.get_folded_operator(wf.num_inactive_orbs, wf.num_active_orbs, wf.num_virtual_orbs)
-    return wf.QI.quantum_expectation_value(H)
-
-
-def orbital_rotation_gradient(
-    placeholder,  # pylint: disable=unused-argument
-    wf,
-) -> np.ndarray:
-    """Calcuate electronic gradient with respect to orbital rotations.
+        Args:
+            parameters: Ansatz and orbital rotation parameters.
+            theta_optimization: Doing theta optimization.
+            kappa_optimization: Doing kappa optimization.
 
-    Args:
-        placeholder: Placeholder for kappa parameters, these are fetched OOP style instead.
-        wf: Wave function object.
+        Returns:
+            Electronic energy.
+        """
+        num_kappa = 0
+        if kappa_optimization:
+            num_kappa = len(self.kappa_idx)
+            self.kappa = parameters[:num_kappa]
+            self._move_cep()
+        if theta_optimization:
+            self.thetas = parameters[num_kappa:]
+        # Build operator
+        if theta_optimization:
+            H = hamiltonian_0i_0a(self.h_mo, self.g_mo, self.num_inactive_orbs, self.num_active_orbs)
+            H = H.get_folded_operator(self.num_inactive_orbs, self.num_active_orbs, self.num_virtual_orbs)
+            return self.QI.quantum_expectation_value(H)
+        rdms = ReducedDenstiyMatrix(
```

Comment:
Why is the energy only calcuated via RDM if theta_optimization = False?

Reply:
Because RDM construction is the same price or more expensive than calculating the Hamiltonian (some of the RDM terms might not be needed for the energy because the integrals are zero).

But if theta_optimization is false, then the RDM's calculated to get the energy will also be used to get the gradients for the orbital optimization, hence it is worth it then.

Reply:
Ok. makes sense. How about a one-liner comment explaining this